argo/src/Argo.hs

```diff
@@ -351,28 +376,70 @@ handleRequest logger respond app req =
     Nothing -> throwIO $ (methodNotFound method) { errorID = reqID }
     Just (Command, m) ->
       withRequestID $
-      do answer <- modifyMVar theState $ runMethod (m params) logger
+      do stateID <- getStateID req
+         answer <- withMVar theState $
+           \contents ->
+             stateRecipe contents stateID >>=
+             \case
+               Nothing ->
+                 throwIO $ unknownStateID stateID
+               Just r ->
+                 do initAppState <- getAppState (rerunStep reqID app logger) contents r
+                    let r' = (method, view requestParams req) : r
+                    let fileReader = freshStateReader contents r'
+                    (newAppState, result) <- runMethod (m params) fileReader logger initAppState
+                    sid' <- newStateID
+                    modifyIORef' (view stateRecipes contents) $
+                      set (at sid') (Just r')
+                    modifyIORef' (view appStateCache contents) $
+                      set (at r') (Just newAppState)
+                    return $ addStateID result sid'
          let response = JSON.object [ "jsonrpc" .= jsonRPCVersion
                                     , "id" .= reqID
                                     , "result" .= answer
                                     ]
          respond (JSON.encode response)
     Just (Query, m) ->
       withRequestID $
-      do (_, answer) <- runMethod (m params) logger =<< readMVar theState
+      do stateID <- getStateID req
+         answer <- withMVar theState $
+           \contents ->
+             stateRecipe contents stateID >>=
+             \case
+               Nothing -> throwIO $ unknownStateID stateID
+               Just r ->
+                 do theAppState <- getAppState (rerunStep reqID app logger) contents r
+                    let fileReader = B.readFile -- No caching of this state
```

Comment:
I'm not sure I understand what's happening here. Why do we cache `Command` state but not `Query` state?

Reply:
I believe that the comment in the newly-pushed version answers this.

Because a query by definition does not modify the state, we do not create a new server-side state, nor do we return a new state ID.